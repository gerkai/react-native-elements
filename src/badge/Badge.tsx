import React from 'react';
import {
  StyleSheet,
  Text,
  View,
  TouchableOpacity,
  TextProps,
  StyleProp,
  ViewStyle,
} from 'react-native';
import { withTheme } from '../config';
import { ThemeProps } from '../config';
import { renderNode } from '../helpers';

export type BadgeProps = {
  containerStyle?: StyleProp<ViewStyle>;
  badgeStyle?: StyleProp<ViewStyle>;
  textProps?: TextProps;
  textStyle?: StyleProp<ViewStyle>;
  value?: React.ReactNode;
  onPress?: (...args: any[]) => any;
  Component?: typeof React.Component;
  status?: 'primary' | 'success' | 'warning' | 'error';
};

const Badge: React.FunctionComponent<BadgeProps & ThemeProps<BadgeProps>> = (
  props
) => {
  const {
    containerStyle,
    textStyle,
    textProps,
    badgeStyle,
    onPress,
    Component = onPress ? TouchableOpacity : View,
    value,
    theme,
    status = 'primary',
    ...attributes
  } = props;
  const element = renderNode(Text, value, {
    style: StyleSheet.flatten([styles.text, textStyle && textStyle]),
    ...textProps,
  });
  return (
    <View style={StyleSheet.flatten([containerStyle && containerStyle])}>
      <Component
        {...attributes}
        style={StyleSheet.flatten([
          {
            alignSelf: 'center',
            minWidth: size,
            height: size,
            borderRadius: size / 2,
            alignItems: 'center',
            justifyContent: 'center',
            backgroundColor: theme.colors[status],
            borderWidth: StyleSheet.hairlineWidth,
            borderColor: '#fff',
          },
          !element && styles.miniBadge,
          badgeStyle && badgeStyle,
        ])}
        onPress={onPress}
      >
        {element}
      </Component>
    </View>
  );
};

const size = 18;
const miniSize = 8;

const styles = StyleSheet.create({
  miniBadge: {
    paddingHorizontal: 0,
    paddingVertical: 0,
    minWidth: miniSize,
    height: miniSize,
    borderRadius: miniSize / 2,
  },
  text: {
    fontSize: 12,
    color: 'white',
    paddingHorizontal: 4,
  },
});

export { Badge };
export default withTheme(Badge, 'Badge');
