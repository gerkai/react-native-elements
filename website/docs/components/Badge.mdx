---
id: badge

title: Badge
---

import Usage from "./usage/Badge.mdx";

import { Badge } from "react-native-elements";

import Tabs from "@theme/Tabs";

import TabItem from "@theme/TabItem";

Badges are small components typically used to communicate a numerical value or indicate the status of an item to the user.

## Usage

<Usage />

## Props

### Badge

- [Component](#component)
- [badgeStyle](#badgestyle)
- [containerStyle](#containerstyle)
- [onLongPress](#onlongpress)
- [onPress](#onpress)
- [onPressIn](#onpressin)
- [onPressOut](#onpressout)
- [pressableProps](#pressableprops)
- [status](#status)
- [textProps](#textprops)
- [textStyle](#textstyle)
- [value](#value)

## Reference

### Badge

#### Component

Custom component to replace the badge outer component.

| Type            | Default                                           |
| --------------- | ------------------------------------------------- |
| React Component | `Press handlers present then Pressable else View` |

---

#### badgeStyle

Additional styling for badge (background) view component.

| Type               | Default |
| ------------------ | ------- |
| View style(Object) | None    |

---

#### containerStyle

Style for the container.

| Type               | Default |
| ------------------ | ------- |
| View style(Object) | None    |

---

#### onLongPress

Called when a long-tap gesture is detected.

| Type                         | Default |
| ---------------------------- | ------- |
| GestureResponderEventHandler | None    |

---

#### onPress

Called when a single tap gesture is detected.

| Type     | Default |
| -------- | ------- |
| Function | None    |

---

#### onPressIn

Called when a touch is engaged before `onPress`.

| Type                         | Default |
| ---------------------------- | ------- |
| GestureResponderEventHandler | None    |

---

#### onPressOut

Called when a touch is released before `onPress`.

| Type                         | Default |
| ---------------------------- | ------- |
| GestureResponderEventHandler | None    |

---

#### pressableProps

| Type                                 | Default |
| ------------------------------------ | ------- |
| PressableProps except click handlers | None    |

---

#### status

Determines color of the indicator.

| Type                                           | Default |
| ---------------------------------------------- | ------- |
| "primary" or "success" or "warning" or "error" | primary |

---

#### textProps

Extra props for text component.

| Type      | Default |
| --------- | ------- |
| TextProps | None    |

---

#### textStyle

Extra styling for icon component.

| Type               | Default |
| ------------------ | ------- |
| Text Style(Object) | None    |

---

#### value

Text value to be displayed by badge, defaults to empty.

| Type            | Default |
| --------------- | ------- |
| React.ReactNode | None    |

---
