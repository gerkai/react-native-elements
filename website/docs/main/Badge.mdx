---
id: badge

title: Badge

slug: /badge
---

import Usage from "./usage/Badge/Badge.mdx";

import { Badge } from "react-native-elements";

import Tabs from "@theme/Tabs";

import TabItem from "@theme/TabItem";

Badges are small components typically used to communicate a numerical value or indicate the status of an item to the user.

## Usage

<Usage />

## Props

### Badge

- [Component](#component)
- [badgeStyle](#badgestyle)
- [containerStyle](#containerstyle)
- [status](#status)
- [textProps](#textprops)
- [textStyle](#textstyle)
- [value](#value)

## Reference

### Badge

#### Component

Custom component to replace the badge outer component.

| Type            | Default                                           |
| --------------- | ------------------------------------------------- |
| React Component | `Press handlers present then Pressable else View` |

---

#### badgeStyle

Additional styling for badge (background) view component.

| Type               | Default |
| ------------------ | ------- |
| View style(Object) | None    |

---

#### containerStyle

Style for the container.

| Type               | Default |
| ------------------ | ------- |
| View style(Object) | None    |

---

#### status

Determines color of the indicator.

| Type                                           | Default |
| ---------------------------------------------- | ------- |
| "primary" or "success" or "warning" or "error" | primary |

---

#### textProps

Extra props for text component.

| Type      | Default |
| --------- | ------- |
| TextProps | None    |

---

#### textStyle

Extra styling for icon component.

| Type               | Default |
| ------------------ | ------- |
| Text Style(Object) | None    |

---

#### value

Text value to be displayed by badge, defaults to empty.

| Type            | Default |
| --------------- | ------- |
| React.ReactNode | None    |

---
