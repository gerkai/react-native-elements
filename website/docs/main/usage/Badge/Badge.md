import Snack from './snack/index.md'

### withBadge Higher-Order Component

The withBadge HOC allows you to easily add badges to icons and other components.

<Snack />
