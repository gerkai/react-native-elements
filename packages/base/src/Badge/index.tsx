import { withTheme } from '../config';
import { Badge, BadgeProps } from './Badge';
import { withBadge } from './withBadge';

export { Badge, withBadge };
export type { BadgeProps };
export default withTheme(Badge, 'Badge');
