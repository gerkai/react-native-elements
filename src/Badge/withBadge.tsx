import React from 'react';
import { StyleProp, StyleSheet, View, ViewStyle } from 'react-native';
import Badge, { BadgeProps } from './index';

type withBadgeOptions = {
  bottom?: number;
  left?: number;
  right?: number;
  top?: number;
  hidden?: boolean;
  containerStyle?: StyleProp<ViewStyle>;
} & BadgeProps;

export const withBadge = (
  value: React.ReactNode | ((props: any) => React.ReactNode),
  options: withBadgeOptions = {}
) => (WrappedComponent: React.ComponentType<any>): React.ComponentType => {
  const WithBadge = (props: any) => {
    const {
      bottom,
      hidden = false,
      left,
      containerStyle,
      ...badgeProps
    } = options;
    let { right = -16, top = -1 } = options;
    if (!value) {
      right = -3;
      top = 3;
    }
    const badgeValue = typeof value === 'function' ? value(props) : value;
    return (
      <View style={StyleSheet.flatten([styles.container, containerStyle])}>
        <WrappedComponent {...props} />

        {!hidden && (
          <Badge
            value={badgeValue}
            status="error"
            containerStyle={StyleSheet.flatten([
              styles.badgeContainer,
              { bottom, left, right, top },
            ])}
            {...badgeProps}
          />
        )}
      </View>
    );
  };
  WithBadge.displayName = `WithBadge(${
    WrappedComponent.displayName || WrappedComponent.name || 'Component'
  })`;
  return WithBadge;
};

const styles = StyleSheet.create({
  badgeContainer: {
    position: 'absolute',
  },
  container: {
    alignItems: 'center',
    justifyContent: 'center',
    position: 'relative',
  },
});
