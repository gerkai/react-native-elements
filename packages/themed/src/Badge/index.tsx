import { withTheme } from '../config';
import {
  Badge,
  BadgeProps,
} from '@react-native-elements/base/dist/Badge/Badge';
import { withBadge } from '@react-native-elements/base/dist/Badge/withBadge';

export { Badge, withBadge };
export type { BadgeProps };
export default withTheme(Badge, 'Badge');
