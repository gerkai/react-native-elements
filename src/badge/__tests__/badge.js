import React from 'react';
import { Text } from 'react-native';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Badge from '../badge';

describe('Badge Component', () => {
  it('should render without issue', () => {
    const component = shallow(<Badge/>);

    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should throw Error if value and child are included', () => {
    expect(() => {
      shallow(<Badge value={"hello"}><Text/></Badge>);
    }).toThrow('Badge can only contain a single child or string value');
  });

  it('should render if element included', () => {
    const component = shallow(<Badge ><Text title='foo' /></Badge>);

    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
    expect(component.props().children.props.title).toBe('foo');
  });

  it('old badge props should still work', () => {
    const component = shallow(<Badge badge={{value: 'foo'}} />);

    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  })
});
