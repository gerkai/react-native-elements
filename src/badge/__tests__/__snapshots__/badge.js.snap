// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge Component should allow adding custom element 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
>
  Hello
</Text>
`;

exports[`Badge Component should allow wrapper style 1`] = `
<Component
  style={
    Array [
      Object {},
    ]
  }
>
  <Component
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#2089dc",
          "borderRadius": 20,
          "justifyContent": "center",
          "padding": 12,
          "paddingBottom": 3,
          "paddingTop": 3,
        },
        undefined,
      ]
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "white",
            "fontSize": 14,
          },
          undefined,
        ]
      }
    />
  </Component>
</Component>
`;

exports[`Badge Component should apply container style in the badge 1`] = `
<Component
  style={
    Array [
      undefined,
    ]
  }
>
  <Component
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#2089dc",
          "borderRadius": 20,
          "justifyContent": "center",
          "padding": 12,
          "paddingBottom": 3,
          "paddingTop": 3,
        },
        Object {
          "backgroundColor": "orange",
        },
      ]
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "white",
            "fontSize": 14,
          },
          undefined,
        ]
      }
    />
  </Component>
</Component>
`;

exports[`Badge Component should apply text style in the badge 1`] = `
<Component
  style={
    Array [
      undefined,
    ]
  }
>
  <Component
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#2089dc",
          "borderRadius": 20,
          "justifyContent": "center",
          "padding": 12,
          "paddingBottom": 3,
          "paddingTop": 3,
        },
        undefined,
      ]
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "white",
            "fontSize": 14,
          },
          Object {
            "color": "orange",
          },
        ]
      }
    />
  </Component>
</Component>
`;

exports[`Badge Component should pass value props should still work 1`] = `
<Component
  style={
    Array [
      undefined,
    ]
  }
>
  <Component
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#2089dc",
          "borderRadius": 20,
          "justifyContent": "center",
          "padding": 12,
          "paddingBottom": 3,
          "paddingTop": 3,
        },
        undefined,
      ]
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "white",
            "fontSize": 14,
          },
          undefined,
        ]
      }
    >
      foo
    </Text>
  </Component>
</Component>
`;

exports[`Badge Component should render if element included 1`] = `
<Component
  style={
    Array [
      undefined,
    ]
  }
>
  <Component
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#2089dc",
          "borderRadius": 20,
          "justifyContent": "center",
          "padding": 12,
          "paddingBottom": 3,
          "paddingTop": 3,
        },
        undefined,
      ]
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      title="foo"
    />
  </Component>
</Component>
`;

exports[`Badge Component should render without issue 1`] = `
<Component
  style={
    Array [
      undefined,
    ]
  }
>
  <Component
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#2089dc",
          "borderRadius": 20,
          "justifyContent": "center",
          "padding": 12,
          "paddingBottom": 3,
          "paddingTop": 3,
        },
        undefined,
      ]
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "white",
            "fontSize": 14,
          },
          undefined,
        ]
      }
    />
  </Component>
</Component>
`;

exports[`Badge Component should show error if value and child are included 1`] = `
<Component
  style={
    Array [
      undefined,
    ]
  }
>
  <Component
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#2089dc",
          "borderRadius": 20,
          "justifyContent": "center",
          "padding": 12,
          "paddingBottom": 3,
          "paddingTop": 3,
        },
        undefined,
      ]
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    />
  </Component>
</Component>
`;
